Add support for new resource and datasource for Managed DevOps Pools service

ajaykn opened the pull request:
<!--  All Submissions -->


## Community Note
<!-- Please leave the community note as is. -->
* Please vote on this PR by adding a :thumbsup: [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original PR to help the community and maintainers prioritize for review
* Please do not leave comments along the lines of "+1", "me too" or "any updates", they generate extra noise for PR followers and do not help prioritize for review


## Description

<!-- Please include a description below with the reason for the PR, what it is doing, what it is trying to accomplish, and anything relevant for a reviewer to know. 

If this is a breaking change for users please detail how it cannot be avoided and why it should be made in a minor version of the provider -->

Managed DevOps Pools is a new Azure Service went to GA. 
This PR adds terraform support.

Public Docs:
https://learn.microsoft.com/en-us/azure/devops/managed-devops-pools/?view=azure-devops

Azure Rest Api Specs:
https://github.com/Azure/azure-rest-api-specs/tree/main/specification/devopsinfrastructure/resource-manager/Microsoft.DevOpsInfrastructure/stable/2025-01-21

closing https://github.com/hashicorp/terraform-provider-azurerm/issues/26412

Also, I am from Managed DevOps Pool team and we are taking over the community PR as per the discussion here
https://github.com/hashicorp/terraform-provider-azurerm/pull/27457#issuecomment-2445106820


## PR Checklist

- [x] I have followed the guidelines in our [Contributing Documentation](../blob/main/contributing/README.md).
- [x] I have checked to ensure there aren't other open [Pull Requests](../pulls) for the same update/change.
- [x] I have checked if my changes close any open issues. If so please include appropriate [closing keywords](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword) below.
- [x] I have updated/added Documentation as required written in a helpful and kind way to assist users that may be unfamiliar with the resource / data source.
- [x] I have used a meaningful PR title to help maintainers and other users understand this change and help prevent duplicate work. 
For example: “`resource_name_here` - description of change e.g. adding property `new_property_name_here`”


<!-- You can erase any parts of this template below this point that are not applicable to your Pull Request. -->


## Changes to existing Resource / Data Source

- [x] I have added an explanation of what my changes do and why I'd like you to include them (This may be covered by linking to an issue above, but may benefit from additional explanation).
- [x] I have written new tests for my resource or datasource changes & updated any relevent documentation.
- [x] I have successfully run tests with my changes locally. If not, please provide details on testing challenges that prevented you running the tests.
- [ ] (For changes that include a **state migration only**). I have manually tested the migration path between relevant versions of the provider.


## Testing 

- [x] My submission includes Test coverage as described in the [Contribution Guide](../blob/main/contributing/topics/guide-new-resource.md) and the tests pass. (if this is not possible for any reason, please include details of why you did or could not add test coverage)

<!-- Please include testing logs or evidence here or an explanation on why no testing evidence can be provided. 

For state migrations please test the changes locally and provide details here, such as the versions involved in testing the migration path. For further details on testing state migration changes please see our guide on [state migrations](https://github.com/hashicorp/terraform-provider-azurerm/blob/main/contributing/topics/guide-state-migrations.md#testing) in the contributor documentation. -->


## Change Log

Below please provide what should go into the changelog (if anything) conforming to the [Changelog Format documented here](../blob/main/contributing/topics/maintainer-changelog.md).

<!-- Replace the changelog example below with your entry. One resource per line. -->

`FEATURES:`

* `azurerm_managed_devops_pool - Add support for new resource [GH-28325]`
* `azurerm_managed_devops_pool - Add support for new data source [GH-28325]`

<!-- What type of PR is this? -->
This is a (please select all that apply):

- [ ] Bug Fix
- [x] New Feature (ie adding a service, resource, or data source)
- [ ] Enhancement
- [ ] Breaking Change


## Related Issue(s)
Closes #26412

Local testing:

1. Datasource
![image](https://github.com/user-attachments/assets/f4b946b9-a1b7-4a9c-8084-1cab8134da1f)

2. Terraform import

![image](https://github.com/user-attachments/assets/84b3691d-3f8a-4c31-8f9d-7d92854b2ccc)

3. Create

![image](https://github.com/user-attachments/assets/5564b992-82c0-4e8d-9db8-42e96dff2d92)

4. Update

![image](https://github.com/user-attachments/assets/dc9f96eb-3e36-4d49-8bc8-efdc6f6a28bd)

5. Delete

![image](https://github.com/user-attachments/assets/1b725fa6-f920-4254-8fc1-993713a96eb3)


> [!NOTE] 
> If this PR changes meaningfully during the course of review please update the title and description as required.

ajaykn committed:
Adding new service package manageddevopspool

ajaykn committed:
make generate

ajaykn committed:
make fmt

ajaykn committed:
Adding pools Client

ajaykn committed:
Adding resource

ajaykn committed:
Adding resource tests base

ajaykn committed:
fix base tests

ajaykn committed:
Fix linting

ajaykn committed:
Fix assignments in tests

ajaykn committed:
Adding github label

ajaykn committed:
fix lint

ajaykn commented:
> Hi @ajaykn We're eagerly awaiting the new Terraform resources - do you have any indication of when these might be released?

@jgschwendswica I will plan to wrap up this soon.  Pending to add few unit tests and bit testing locally.

ajaykn committed:
Adding datasource and tests partially

ajaykn committed:
fix errors

ajaykn committed:
lint error

ajaykn committed:
Fix Agent profile

ajaykn committed:
remove unwanted log

TyranVDM commented:
@ajaykn any rough timelines for this?

jsloan117 commented:
Our company is very interested in utilizing this new feature as well!

tylerforaie commented:
It would awesome to see this get implemented soon

ajaykn committed:
Flatten back Read resource to model

ajaykn committed:
Handling resource predictions

ajaykn committed:
handling data source with schema updates and assignments

ajaykn committed:
Add provsioning_state to the response

ajaykn committed:
add missing import

ajaykn committed:
terrafmt on test files

ajaykn committed:
fmt and lint-ci fixes

ajaykn committed:
upgrade to latest

ajaykn committed:
add missing reference

ajaykn committed:
fix deps and add certificateStoreName support

ajaykn committed:
fix pointers nil refs

ajaykn committed:
Adding documentation

ajaykn committed:
add docs for datasource and validation fns

ajaykn commented:
> @ajaykn any rough timelines for this?

This is ready for review, will followup on reviews and get it released.

ajaykn commented:
> @stephybun Can I please get some feedback on this

@stephybun Can we please get reviews on this.

ms-henglu commented on internal/services/manageddevopspools/managed_devops_pool_resource_test.go:
```diff
@@ -0,0 +1,266 @@
+  fabric_profile {
+    kind = "Vmss"
+    images {
+      resource_id = "/Subscriptions/e21f7bce-1728-44ad-a62d-344064a0d69a/Providers/Microsoft.Compute/Locations/australiaeast/publishers/canonical/artifacttypes/vmimage/offers/0001-com-ubuntu-server-focal/skus/20_04-lts-gen2/versions/latest"
```
The testcases will run in the Hashicorp's testing subscriptions, can this resource will be used?

ajaykn replied:
updated to use the data.client().subscriptoin and these are compute images from marketplace available on every subscription.

ms-henglu commented on internal/services/manageddevopspools/schema_definitions.go:
```diff
@@ -0,0 +1,469 @@
+					Type:     pluginsdk.TypeString,
+					Optional: true,
+				},
+				"days_data": {
```
From its doc, this field only accepts `json` format, maybe add a JSON string validation function?

ms-henglu commented on internal/services/manageddevopspools/schema_definitions.go:
```diff
@@ -0,0 +1,469 @@
+						Schema: map[string]*pluginsdk.Schema{
+							"subnet_id": {
+								Type:     pluginsdk.TypeString,
+								Required: true,
```
Please add validation function for subnet id.

ajaykn replied:
added, please check

ms-henglu commented on internal/services/manageddevopspools/schema_definitions.go:
```diff
@@ -0,0 +1,469 @@
+					},
+				},
+				"os_profile": OsProfileSchema(),
+				"sku": {
```
Can this be simplified to `sku_name`(string, required)?

ajaykn replied:
This is for future extension, where we might see Sku :{name, tier, cpu, ram)

ajaykn commented on internal/services/manageddevopspools/managed_devops_pool_resource.go:
```diff
@@ -0,0 +1,239 @@
+
+func (ManagedDevOpsPoolResource) Attributes() map[string]*pluginsdk.Schema {
+	return map[string]*pluginsdk.Schema{
+		"type": {
```
What if customers want to check the provisioning_state in their flow ?

ajaykn committed:
Adding validations and addressing review comments

ajaykn commented on internal/services/manageddevopspools/managed_devops_pool_data_source.go:
```diff
@@ -0,0 +1,133 @@
+			Computed: true,
+		},
+		"tags": commonschema.TagsDataSource(),
+		"type": {
```
Removed `type` but kept `provisioning_state` so customers can see the status of data source

ms-henglu replied:
Please also remove them from the data model.

ajaykn replied:
Good catch, removed from model and assignments as well in resource, data source.

ajaykn committed:
order elements in resource schema

ajaykn commented on internal/services/manageddevopspools/managed_devops_pool_resource.go:
```diff
@@ -0,0 +1,239 @@
+
+func (ManagedDevOpsPoolResource) Arguments() map[string]*pluginsdk.Schema {
+	return map[string]*pluginsdk.Schema{
+		"agent_profile": AgentProfileSchema(),
```
Thanks for guidance. Updated, please check

ajaykn committed:
removing type fro  models and implementation as well

ajaykn committed:
Address review comments

ajaykn commented on internal/services/manageddevopspools/managed_devops_pool_resource_test.go:
```diff
@@ -0,0 +1,266 @@
+  features {}
+}
+
+esource "azurerm_managed_devops_pool" "test" {
```
Thanks a catching this

ajaykn commented on internal/services/manageddevopspools/managed_devops_pool_resource.go:
```diff
@@ -0,0 +1,235 @@
+
+			properties := existing.Model
+
+			if err := expandResourceModel(config, properties); err != nil {
```
@ms-zhenhua  In our case, we support PUT operation. we dont support PATCH, so we need to pass anyways the entire object.
Can you please share even in this case do we need to do `metadata.ResourceData.HasChange` ?

ms-zhenhua replied:
For whatever PUT or PATCH API, in the Update method, only changed properties should be assigned a new value. For those unchanged properties, they should reuse values inherited from GET method.

ajaykn replied:
Got it. I have followed your suggestion and updated. please check.

ajaykn commented on internal/services/manageddevopspools/managed_devops_pool_resource.go:
```diff
@@ -0,0 +1,235 @@
+			}
+
+			var payload pools.Pool
+			if err := expandResourceModel(config, &payload); err != nil {
```
What does inline mean in this context. Can you please share more details ?

ms-zhenhua replied:
it means directly put code here without define `expandResourceModel` method. You may refer to other existing resource code for more details.

ms-zhenhua replied:
we should keep the same code style with existing resources

ajaykn replied:
Done updated, please check

ajaykn committed:
fix more review comments and update docs

ajaykn committed:
fmt fix

ajaykn committed:
missing import

ajaykn committed:
tflint fix

ms-zhenhua commented on internal/services/manageddevopspools/managed_devops_pool_resource.go:
```diff
@@ -0,0 +1,242 @@
+			ForceNew: true,
+			ValidateFunc: validation.StringMatch(
+				regexp.MustCompile(`^[a-zA-Z0-9][a-zA-Z0-9-.]*[a-zA-Z0-9-]$`),
+				"Name must consist of alphanumeric characters, '.' or '-'. Name must also start with alphanumeric characters and cannot end with '.'.",
```
```suggestion
				"`name` can only include alphanumeric characters, periods (.) and hyphens (-). It must also start with alphanumeric characters and cannot end with periods (.).",
```

ms-zhenhua commented on internal/services/manageddevopspools/managed_devops_pool_resource.go:
```diff
@@ -0,0 +1,242 @@
+		},
+		"location":      commonschema.Location(),
+		"agent_profile": AgentProfileSchema(),
+		"dev_center_project_resource_id": {
```
could you reuse `dev_center_project_id` in `dev_center_project_pool_resource.go` ?

ajaykn replied:
Good suggestion. I have updated., please check

ms-zhenhua commented on website/docs/r/managed_devops_pool.html.markdown:
```diff
@@ -0,0 +1,292 @@
+
+A `secrets_management_settings` block supports the following:
+
+* `certificate_store_location` - (Optional) Specified where to store certificates on the machine.
```
```suggestion
* `certificate_store_location` - (Optional) Specifies where to store certificates on the machine.
```

ajaykn committed:
Fixing review comment on Devcenter project, name validations

ajaykn committed:
Expand inline and update if changed only

ajaykn committed:
build fixes

ajaykn committed:
fix the return types of expand methods

ajaykn committed:
Merge branch 'main' into ajaykn/support-devopsinfrastructure

ajaykn committed:
Fixing the deventer package version in import

ajaykn committed:
make fmt

ajaykn committed:
Update resource group name schema with common schema

ajaykn commented on internal/services/manageddevopspools/managed_devops_pool_resource.go:
```diff
@@ -0,0 +1,235 @@
+		},
+		"location":      commonschema.Location(),
+		"agent_profile": AgentProfileSchema(),
+		"dev_center_project_resource_id": {
```
Done, thanks for suggestion

ajaykn committed:
nit in docs

ms-zhenhua commented on internal/services/manageddevopspools/managed_devops_pool_resource.go:
```diff
@@ -140,13 +159,57 @@ func (r ManagedDevOpsPoolResource) Update() sdk.ResourceFunc {
+				payload.Properties.MaximumConcurrency = config.MaximumConcurrency
+			}
+
+			if metadata.ResourceData.HasChange("location") {
```
`Location` is a `ForceNew` property which will not enter `Update` method. Please double check the code and remove all `ForceNew` properties from `Update`

ajaykn replied:
I see, thanks for sharing. Have updated and cross checked all ForceNew properties (location and ResourceGroup) and now none of them are being used in Update flow.

ajaykn committed:
fix location in update flow

ms-henglu reviewed (dismissed):
Thanks for the PR, LGTM

juicybaba commented:
merge, merge, merge :wink:

katbyte commented on website/docs/r/managed_devops_pool.html.markdown:
```diff
@@ -0,0 +1,292 @@
+
+* `kind` - (Required) Defines the type of agent profile. Accepted values are: `Stateful` and `Stateless`.
+
+* `grace_period_time_span` - (Optional) How long should the machine be kept around after it ran a workload when there are no stand-by agents. The maximum is one week. This is needed when kind is `Stateful`.
```
what format is this in? seconds? iso duration?
```suggestion
* `grace_period_in_seconds` - (Optional) How long should the machine be kept around after it ran a workload when there are no stand-by agents. The maximum is one week. This is needed when kind is `Stateful`.
```
or 
```suggestion
* `grace_period_Duration` - (Optional) How long should the machine be kept around after it ran a workload when there are no stand-by agents in ISO8601 duration format, for example `P14D`. The maximum is one week. This is required when kind is `Stateful`.
```

ajaykn replied:
This is our public docs for our reference and want to keep it in the same way.
However, I have updated the description. please check.

https://learn.microsoft.com/en-us/azure/devops/managed-devops-pools/configure-scaling?view=azure-devops&tabs=azure-portal
![image](https://github.com/user-attachments/assets/6822928b-e899-49a2-8544-a936f4f9ba0c)

katbyte commented on website/docs/r/managed_devops_pool.html.markdown:
```diff
@@ -0,0 +1,292 @@
+
+* `grace_period_time_span` - (Optional) How long should the machine be kept around after it ran a workload when there are no stand-by agents. The maximum is one week. This is needed when kind is `Stateful`.
+
+* `max_agent_lifetime` - (Optional) How long should stateful machines be kept around. The maximum is one week. This is needed when kind is `Stateful`.
```
same as above

katbyte commented on internal/services/manageddevopspools/managed_devops_pool_resource.go:
```diff
@@ -0,0 +1,301 @@
+		},
+		"resource_group_name":            commonschema.ResourceGroupName(),
+		"location":                       commonschema.Location(),
+		"agent_profile":                  AgentProfileSchema(),
```
we do not abstract schema like this, could you please inline it and get rid of this function?

ajaykn replied:
Curious to know the reson behind this.
I made it separate, just to avoid a very large resource.go file and keep all schema defn's in one place.
However went ahead and made changes according to your suggesttion.

Do we need to do same for datasource.go as well ?

katbyte commented on internal/services/manageddevopspools/managed_devops_pool_resource.go:
```diff
@@ -0,0 +1,301 @@
+		"location":                       commonschema.Location(),
+		"agent_profile":                  AgentProfileSchema(),
+		"dev_center_project_resource_id": commonschema.ResourceIDReferenceRequired(&projects.ProjectId{}),
+		"fabric_profile":                 FabricProfileSchema(),
```
and here

katbyte commented on internal/services/manageddevopspools/managed_devops_pool_resource.go:
```diff
@@ -0,0 +1,301 @@
+			Type:     pluginsdk.TypeInt,
+			Required: true,
+		},
+		"organization_profile": OrganizationProfileSchema(),
```
and here

ajaykn replied:
done

katbyte reviewed (dismissed):
Thanks @ajaykn - i've taken a quick look over this and left some comments inline. once those are addressed we can give this a more through review

ajaykn committed:
Addressing review comments ot inline schema and docs fixes

ajaykn committed:
make fmt

sreallymatt commented on website/docs/r/managed_devops_pool.html.markdown:
```diff
@@ -0,0 +1,292 @@
+
+* `provisioning_state` - The status of the pool operation.
+
+* `type` - This is the resource type `Microsoft.DevOpsInfrastructure/pools`.
```
This isn't in the schema

sreallymatt reviewed (changes requested):
Hi @ajaykn, I've left some comments inline. Some are suggestions, others are questions on how this API behaves as that will influence the schema design for this resource.

ajaykn commented on internal/services/manageddevopspools/managed_devops_pool_resource.go:
```diff
@@ -0,0 +1,444 @@
+								"subnet_id": {
+									Type:     pluginsdk.TypeString,
+									Required: true,
+									ValidateFunc: validation.StringMatch(
+										regexp.MustCompile(`^/subscriptions/[0-9a-fA-F-]{36}/resourceGroups/[-\w._()]+/providers/Microsoft\.Network/virtualNetworks/[-\w._()]+/subnets/[-\w._()]+$`),
+										"Subnet ID must match the format '/subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.Network/virtualNetworks/{vnetName}/subnets/{subnetName}'.",
```
Thanks for sharing, this is very convenient.

ajaykn commented on internal/services/manageddevopspools/managed_devops_pool_data_source.go:
```diff
@@ -0,0 +1,125 @@
+				state.Location = location.Normalize(model.Location)
+				state.Tags = pointer.From(model.Tags)
+
+				if modelIdentity := model.Identity; modelIdentity != nil {
+					identity, err := identity.FlattenSystemAndUserAssignedMapToModel(pointer.To((identity.SystemAndUserAssignedMap)(pointer.From(model.Identity))))
+					if err != nil {
+						return err
+					}
+
+					state.Identity = pointer.From(identity)
+				}
```
I see. thanks for sharing , this is very convenient.

ajaykn commented on internal/services/manageddevopspools/managed_devops_pool_data_source.go:
```diff
@@ -0,0 +1,125 @@
+
+func (ManagedDevOpsPoolDataSource) Attributes() map[string]*pluginsdk.Schema {
+	return map[string]*pluginsdk.Schema{
+		"agent_profile": AgentProfileComputedSchema(),
```
Got it, I have moved. Thanks for suggestion

ajaykn commented on internal/services/manageddevopspools/managed_devops_pool_resource.go:
```diff
@@ -0,0 +1,444 @@
+										Type: pluginsdk.TypeString,
+									},
+								},
+								"kind": {
+									Type:         pluginsdk.TypeString,
+									Required:     true,
+									ValidateFunc: validation.StringInSlice(pools.PossibleValuesForAzureDevOpsPermissionType(), false),
```
yes `groups` and `users` are for `SpecificAccounts` only. I will make change accordingly

ajaykn commented on internal/services/manageddevopspools/managed_devops_pool_resource_test.go:
```diff
@@ -0,0 +1,266 @@
+  fabric_profile {
+    kind = "Vmss"
+    images {
+      resource_id = "/Subscriptions/%s/Providers/Microsoft.Compute/Locations/australiaeast/publishers/canonical/artifacttypes/vmimage/offers/0001-com-ubuntu-server-focal/skus/20_04-lts-gen2/versions/latest"
```
This is nice,  have updated in r.template part.

ajaykn commented on internal/services/manageddevopspools/schema_definitions.go:
```diff
@@ -0,0 +1,326 @@
+						`Buffer must be "*" or value between 1 and 100.`,
+					),
+				},
+				"resource_id": {
+					Type:     pluginsdk.TypeString,
+					Optional: true,
+				},
+				"well_known_image_name": {
+					Type:     pluginsdk.TypeString,
+					Optional: true,
+				},
```
Thats right, only one of this needed. added the `OneOf` validation

ajaykn commented on internal/services/manageddevopspools/schema_definitions.go:
```diff
@@ -0,0 +1,326 @@
+					Optional:     true,
+					ValidateFunc: validation.StringInSlice(pools.PossibleValuesForCertificateStoreNameOption(), false),
+				},
+				"key_exportable": {
```
I wanted to keep the variables close to API, so customers who are using the terraform resource wont have confusion.
But having said that, I understand the guidelines and naming patterns in terraform.
Will go ahead and change as per your suggestion only in the exposed schema, but underlying variables will be `KeyExportable`, so its easy to corelate in expands and flattens.

ajaykn commented on website/docs/d/managed_devops_pool.html.markdown:
```diff
@@ -0,0 +1,216 @@
+
+* `tags` - A mapping of tags assigned to the Managed DevOps Pool.
+
+* `type` - This is the resource type `Microsoft.DevOpsInfrastructure/pools.
```
good catch. I have cleaned up before in schema but missed in docs. Now updated.

ajaykn commented on website/docs/d/managed_devops_pool.html.markdown:
```diff
@@ -0,0 +1,216 @@
+
+---
+
+An `agent_profile` block supports the following:
```
Thanks for this suggestion

ajaykn commented on internal/services/manageddevopspools/helpers.go:
```diff
@@ -0,0 +1,500 @@
+	}
+
+	agentProfile := input[0]
+	resource_predictions := expandResourcePredictionsModel(agentProfile.ResourcePredictions)
```
Oh, thanks for catching this. fixe.d

ajaykn commented on internal/services/manageddevopspools/schema_definitions.go:
```diff
@@ -0,0 +1,326 @@
+					Required:     true,
+					ValidateFunc: validation.StringInSlice(pools.PossibleValuesForLogonType(), false),
+				},
+				"secrets_management_settings": SecretsManagementSettingsSchema(),
```
Makes sense, thanks for the link.

ajaykn commented on website/docs/d/managed_devops_pool.html.markdown:
```diff
@@ -0,0 +1,216 @@
+
+* `drive_letter` - The drive letter for the empty data disk. If not specified, it will be the first available letter.
+
+* `storage_account_type` - The storage Account type to be used for the data disk. If omitted, the default is "standard_lrs".
```
Thanks for suggestion: Updated to: `The storage account type of the data disk.`

ajaykn committed:
Addressing few review comments in docs, schemas inline

ajaykn committed:
make fmt

sreallymatt commented on internal/services/manageddevopspools/schema_definitions.go:
```diff
@@ -0,0 +1,326 @@
+	}
+}
+
+func AgentProfileComputedSchema() *pluginsdk.Schema {
```
Yes, we'd duplicate them and mark one as computed. As an example, [container apps](https://github.com/hashicorp/terraform-provider-azurerm/blob/main/internal/services/containerapps/helpers/container_apps.go#L234)

QixiaLu replied:
Updated schema in data source to be computed

sreallymatt commented on website/docs/d/managed_devops_pool.html.markdown:
```diff
@@ -0,0 +1,216 @@
+
+* `kind` - Determines how the stand-by scheme should be provided. Possible values are: `Manual` and `Automatic`.
+
+* `prediction_preference` - Specifies the desired balance between cost and performance. Accepted values are: `MostCostEffective`, `MoreCostEffective`, `Balanced`, `MorePerformance`, and `BestPerformance`.
```
Correct, since a data source just retrieves data, the user doesn't really need to know all possible values. For resources, we add this so that the user knows what the provider will accept.

QixiaLu commented:
Hi @ajaykn , thanks for contributing to AzureRM provider and developing this resource! Just wondering if this PR is still in progress? Cause there are some Github issues requesting for it.

Sharmanand commented:
hi @ajaykn is it ready to shipped?

ajaykn commented:
> Hi @ajaykn , thanks for contributing to AzureRM provider and developing this resource! Could I kindly ask if this PR is still under development? Cause there are some Github issues requesting for it.

As we discussed offline, I was away for vacation in last month and got into oncall support. 
For others, @QixiaLu will be contributing forward and addressing the comments. Thank you so much.

QixiaLu committed:
Address PR comments

- Separate agent profile to stateful and stateless
- Separate fabric profile to vmss_fabric_profile
- Split organization to azure_devops_organization
- Split resource predictions profile to manual and automatic
- Remove permission profile kind property to put rest of into top level
- Inline schemas
- Skip tests
- Update default values
- Update Cumputed field
- Fix documents
- Update identity to user assigned only

Jingwei-MS commented:
I have passed this PR to our engineering folks for the internal review. Thanks!

QixiaLu committed:
Merge branch 'main' of github.com:hashicorp/terraform-provider-azurerm into ajaykn/support-devopsinfrastructure

QixiaLu committed:
Add separate workaround for identity block

QixiaLu committed:
Linter fix

QixiaLu committed:
Fix document

QixiaLu committed:
Fix typo

Lingkar commented on website/docs/r/managed_devops_pool.html.markdown:
```diff
@@ -0,0 +1,314 @@
+
+-> **Note:** More information about supported images can be found in [Microsoft Learn documentation](https://learn.microsoft.com/en-us/azure/devops/managed-devops-pools/configure-pool-settings?view=azure-devops&tabs=azure-portal#images)
+
+-> **Note:** Exactly one of `resource_id` and `well_known_image_name` are required per `image`
```
"An image can only have one or the other."

I think "and" should become "or" for clarity.

magodo replied:
```suggestion
~> **Note:** Exactly one of `resource_id` and `well_known_image_name` are required per `image`
```

Lingkar commented on internal/services/manageddevopspools/managed_devops_pool_resource_test.go:
```diff
@@ -0,0 +1,386 @@
+
+  vmss_fabric_profile {
+    image {
+      resource_id = data.azurerm_platform_image.test.id
```
Would be good to add a test case for adding a vm image via a well_known_image_name?
I am having some issues getting it to work, and would love to see a working example.

QixiaLu commented on internal/services/manageddevopspools/managed_devops_pool_resource_test.go:
```diff
@@ -0,0 +1,386 @@
+
+  vmss_fabric_profile {
+    image {
+      resource_id = data.azurerm_platform_image.test.id
```
sure, i'll add a test case for this

QixiaLu replied:
There seemed to be a validation on the API that when `resource_id=""` and `well_known_image_name=*`, it still tries to view this image as from the resource, not from the well known image. I have added a workaround to fix it. And updated the test file to cover this property.

QixiaLu committed:
Update doc, fix validation and include complete test for image

magodo commented on website/docs/r/managed_devops_pool.html.markdown:
```diff
@@ -0,0 +1,314 @@
+    }
+
+    network_profile {
+      subnet_id = "/subscriptions/00000000-0000-0000-0000-000000000000/resourceGroups/group1/providers/Microsoft.Network/virtualNetworks/vnet1/subnets/subnet1"
```
Could we make this a reference to another `azurerm_subnet` resource defined above?

QixiaLu replied:
I have added a `azurerm_subnet` block and reference id from that resource. thanks!

magodo commented on website/docs/r/managed_devops_pool.html.markdown:
```diff
@@ -0,0 +1,314 @@
+
+The following arguments are supported:
+
+* `dev_center_project_resource_id` - (Required) The ID of the Dev Center project.
```
Can we make sure the `name`, `resource_group_name` and `location` are positioned first, to be consistent with the other documents?

magodo commented on website/docs/r/managed_devops_pool.html.markdown:
```diff
@@ -0,0 +1,314 @@
+
+* `stateful_agent_profile` - (Optional) A `stateful_agent_profile` block as defined below. Exactly one of `stateful_agent_profile` or `stateless_agent_profile` must be specified.
+
+* `stateless_agent_profile` - (Optional) A `stateless_agent_profile` block as defined below. Exactly one of `stateful_agent_profile` or `stateless_agent_profile` must be specified.
```
Based on https://github.com/hashicorp/terraform-provider-azurerm/blob/main/contributing/topics/reference-documentation-standards.md#warning-note, this shall be modified to:


```suggestion
* `stateful_agent_profile` - (Optional) A `stateful_agent_profile` block as defined below.

* `stateless_agent_profile` - (Optional) A `stateless_agent_profile` block as defined below.

~> **NOTE:** Exactly one of `stateful_agent_profile` or `stateless_agent_profile` must be specified.
```

magodo commented on website/docs/r/managed_devops_pool.html.markdown:
```diff
@@ -0,0 +1,314 @@
+
+* `manual_resource_predictions_profile` - (Optional) A `manual_resource_predictions_profile` block as defined below. Exactly one of `manual_resource_predictions_profile` or `automatic_resource_predictions_profile` may be specified.
+
+* `automatic_resource_predictions_profile` - (Optional) An `automatic_resource_predictions_profile` block as defined below. Exactly one of `manual_resource_predictions_profile` or `automatic_resource_predictions_profile` may be specified.
```
Same here

magodo commented on website/docs/r/managed_devops_pool.html.markdown:
```diff
@@ -0,0 +1,314 @@
+    }
+
+    network_profile {
+      subnet_id = "/subscriptions/00000000-0000-0000-0000-000000000000/resourceGroups/group1/providers/Microsoft.Network/virtualNetworks/vnet1/subnets/subnet1"
```
Can we reference a `azurerm_subnet` (created above) instead of using a hardcoded resource id?

QixiaLu replied:
I think these comments are duplicated for some reasons. But I have updated it as above

magodo commented on website/docs/r/managed_devops_pool.html.markdown:
```diff
@@ -0,0 +1,314 @@
+
+* `stateful_agent_profile` - (Optional) A `stateful_agent_profile` block as defined below. Exactly one of `stateful_agent_profile` or `stateless_agent_profile` must be specified.
+
+* `stateless_agent_profile` - (Optional) A `stateless_agent_profile` block as defined below. Exactly one of `stateful_agent_profile` or `stateless_agent_profile` must be specified.
```
Based on https://github.com/hashicorp/terraform-provider-azurerm/blob/main/contributing/topics/reference-documentation-standards.md#warning-note, this shall be:


```suggestion
* `stateful_agent_profile` - (Optional) A `stateful_agent_profile` block as defined below.

* `stateless_agent_profile` - (Optional) A `stateless_agent_profile` block as defined below.

~> **NOTE:** Exactly one of `stateful_agent_profile` or `stateless_agent_profile` must be specified.
```

QixiaLu replied:
Updated as above

magodo commented on website/docs/r/managed_devops_pool.html.markdown:
```diff
@@ -0,0 +1,314 @@
+
+* `manual_resource_predictions_profile` - (Optional) A `manual_resource_predictions_profile` block as defined below. Exactly one of `manual_resource_predictions_profile` or `automatic_resource_predictions_profile` may be specified.
+
+* `automatic_resource_predictions_profile` - (Optional) An `automatic_resource_predictions_profile` block as defined below. Exactly one of `manual_resource_predictions_profile` or `automatic_resource_predictions_profile` may be specified.
```
Same here.

magodo reviewed (changes requested):
@QixiaLu I've put some comments below, appreciate if you can take a look! Meanwhile, there are quite a few comments pending resolved in Matt's review: https://github.com/hashicorp/terraform-provider-azurerm/pull/28325#pullrequestreview-2940827094.

QixiaLu commented on internal/services/manageddevopspools/managed_devops_pool_resource.go:
```diff
@@ -0,0 +1,444 @@
+				return metadata.ResourceRequiresImport(r.ResourceType(), id)
+			}
+
+			identity, err := identity.ExpandLegacySystemAndUserAssignedMapFromModel(config.Identity)
```
Updated to be expandedIdentity

QixiaLu commented on internal/services/manageddevopspools/schema_definitions.go:
```diff
@@ -0,0 +1,326 @@
+		MaxItems: 1,
+		Elem: &pluginsdk.Resource{
+			Schema: map[string]*pluginsdk.Schema{
+				"time_zone": {
```
Added default value

QixiaLu commented on internal/services/manageddevopspools/schema_definitions.go:
```diff
@@ -0,0 +1,326 @@
+		MaxItems: 1,
+		Elem: &pluginsdk.Resource{
+			Schema: map[string]*pluginsdk.Schema{
+				"kind": {
```
This documentation contains a more comprehensive explanation of `agent_profile` and how `prediction_profile` works . (https://learn.microsoft.com/en-us/azure/devops/managed-devops-pools/configure-scaling?view=azure-devops&tabs=azure-portal)

Here is a summary of the bahavior based on the above and Azure Portal:

`agent_profile` == `stateful`: 
   (Users can update)
   1.1. grace_period_time_span
   1.2. max_agent_lifetime
   1.3. manual_resource_predictions_profile
   1.4 automatic_resource_predictions_profile

`agent_profile` == `stateless`:
   (Users can update)
   2.1. manual_resource_predictions_profile
   2.2. automatic_resource_predictions_profile
   
 `resource_predictions_profile` == `manual`:
    (Users can update)
    3.1 `resource_predictions`
 
 `resource_predictions_profile` == `automatic`:
   (Users can update)
   4.1 `prediction_prefrence`
   
 So, I have removed the `kind` here, and split `resource_predictions_profile` to `manual_*` and `automatic_*`

QixiaLu commented on internal/services/manageddevopspools/schema_definitions.go:
```diff
@@ -0,0 +1,326 @@
+					ValidateFunc: validation.StringIsNotEmpty,
+				},
+				"resource_predictions":         ResourcePredictionsSchema(),
+				"resource_predictions_profile": ResourcePredictionsProfileSchema(),
```
Updated to be computed for data source

QixiaLu commented on website/docs/d/managed_devops_pool.html.markdown:
```diff
@@ -0,0 +1,216 @@
+
+* `data_disks` - One or more `data_disks` blocks as defined above.
+
+* `os_disk_storage_account_type` - The storage account type of the OS disk. Possible values are: `Premium`, `Standard` and `StandardSSD`.
```
Removed possible values for data source doc

QixiaLu commented on website/docs/d/managed_devops_pool.html.markdown:
```diff
@@ -0,0 +1,216 @@
+
+* `tenant_id` - The tenant ID for the identity.
+
+* `type` - The type of managed service identity. The only possible value is `UserAssigned`.
```
Updated identity block

QixiaLu commented on website/docs/d/managed_devops_pool.html.markdown:
```diff
@@ -0,0 +1,216 @@
+
+* `parallelism` - (Optional) Specifies how many machines can be created at maximum in this organization out of the `maximum_concurrency` of the pool.
+
+* `projects` - (Optional) List of projects in which the pool should be created.
```
Removed `Required/Optional` in data source documentation. And rephrased these lines

QixiaLu commented on website/docs/r/managed_devops_pool.html.markdown:
```diff
@@ -0,0 +1,292 @@
+
+* `drive_letter` - (Optional) The drive letter for the empty data disk. If not specified, it will be the first available letter.
+
+* `storage_account_type` - (Optional) The storage Account type to be used for the data disk. If omitted, the default is "standard_lrs".
```
Added Possible values and Default value to `storage_account_type`

QixiaLu commented on website/docs/r/managed_devops_pool.html.markdown:
```diff
@@ -0,0 +1,292 @@
+
+An `os_profile` block supports the following:
+
+* `logon_type` - (Required) Determines how the service should be run. Accepted values are: `Interactive` and `Service`.
```
Updated to `Possible values are`

QixiaLu committed:
Update doc from review

QixiaLu committed:
Update docs

QixiaLu commented on website/docs/r/managed_devops_pool.html.markdown:
```diff
@@ -0,0 +1,314 @@
+    }
+
+    image {
+      resource_id = "/Subscriptions/00000000-0000-0000-0000-000000000000/Providers/Microsoft.Compute/Locations/australiaeast/publishers/canonical/artifacttypes/vmimage/offers/0001-com-ubuntu-server-focal/skus/20_04-lts-gen2/versions/latest"
```
Added an `azurerm_platform_image` data source, and reference it from there

QixiaLu commented on website/docs/r/managed_devops_pool.html.markdown:
```diff
@@ -0,0 +1,314 @@
+## Arguments Reference
+
+The following arguments are supported:
+
```
Updated

QixiaLu commented on website/docs/r/managed_devops_pool.html.markdown:
```diff
@@ -0,0 +1,314 @@
+
+* `manual_resource_predictions_profile` - (Optional) A `manual_resource_predictions_profile` block as defined below. Exactly one of `manual_resource_predictions_profile` or `automatic_resource_predictions_profile` may be specified.
+
+* `automatic_resource_predictions_profile` - (Optional) An `automatic_resource_predictions_profile` block as defined below. Exactly one of `manual_resource_predictions_profile` or `automatic_resource_predictions_profile` may be specified.
```
Updated as above

QixiaLu committed:
fmt and flatten sku

QixiaLu committed:
Flatten manual resource predictions

QixiaLu committed:
Remodel days_data into TF schema

QixiaLu committed:
Add customizeDiff to validate image

QixiaLu committed:
Add validation for agent count

QixiaLu committed:
Updates permission profile

QixiaLu committed:
make fmt

QixiaLu committed:
fmt and test fixes

QixiaLu commented on website/docs/r/managed_devops_pool.html.markdown:
```diff
@@ -0,0 +1,314 @@
+
+* `image` - (Required) One or more `image` blocks as defined below.
+
+* `sku` - (Required) A `sku` block as defined below.
```
Good point! I have flattened it into `sku_name`. Thanks!

QixiaLu committed:
Update sku_name in data source

magodo commented on internal/services/manageddevopspools/managed_devops_pool_resource.go:
```diff
@@ -0,0 +1,679 @@
+						Elem: &pluginsdk.Resource{
+							Schema: map[string]*pluginsdk.Schema{
+								"aliases": {
+									Type:     pluginsdk.TypeSet,
```
Any reason to use `TypeSet` here? Can we either test or confirm with the service team that whether the order of `aliases` will be conserved, in which case we shall prefer using `TypeList`?

magodo commented on internal/services/manageddevopspools/managed_devops_pool_resource.go:
```diff
@@ -0,0 +1,679 @@
+											},
+											"drive_letter": {
+												Type:     pluginsdk.TypeString,
+												Optional: true,
```
Can we add an validation here?

QixiaLu replied:
Added a validation to make sure it's a single letter and not A, C, D or E. ([reference](https://learn.microsoft.com/en-us/azure/devops/managed-devops-pools/configure-storage?view=azure-devops&tabs=bicep%2Cwindows#attach-an-empty-data-disk))

magodo commented on internal/services/manageddevopspools/managed_devops_pool_resource.go:
```diff
@@ -0,0 +1,679 @@
+												Type:     pluginsdk.TypeSet,
+												Optional: true,
+												Elem: &pluginsdk.Schema{
+													Type: pluginsdk.TypeString,
```
Can we add a validation here?

QixiaLu replied:
Added a validation for email, ([reference](https://learn.microsoft.com/en-us/rest/api/devopspools/pools/create-or-update?view=rest-devopspools-2025-01-21&tabs=HTTP#azuredevopspermissionprofile))

magodo commented on internal/services/manageddevopspools/managed_devops_pool_resource.go:
```diff
@@ -0,0 +1,679 @@
+												Optional: true,
+												Elem: &pluginsdk.Schema{
+													Type: pluginsdk.TypeString,
+												},
```
Can we add a validation here?

QixiaLu replied:
added missing validation for emails

magodo commented on internal/services/manageddevopspools/managed_devops_pool_resource.go:
```diff
@@ -0,0 +1,679 @@
+
+			payload := pools.Pool{
+				Name:     pointer.To(config.Name),
+				Location: location.Normalize(config.Location),
```
We don't have to use the `location.Normalize` in CU, but shall use it in the R. Just FYI.

QixiaLu replied:
thanks! i have removed it from Create

magodo commented on internal/services/manageddevopspools/managed_devops_pool_resource.go:
```diff
@@ -0,0 +1,679 @@
+
+			for _, orgProfile := range model.AzureDevOpsOrganizationProfile {
+				for _, permProfile := range orgProfile.PermissionProfile {
+					if permProfile.Kind != "SpecificAccounts" {
```
Can we use the SDK enum instead of hard coded here?

magodo commented on internal/services/manageddevopspools/managed_devops_pool_resource.go:
```diff
@@ -0,0 +1,679 @@
+				for _, permProfile := range orgProfile.PermissionProfile {
+					if permProfile.Kind != "SpecificAccounts" {
+						if len(permProfile.AdministratorAccounts) > 0 {
+							return fmt.Errorf("administrator_accounts block is not required when permission_profile kind is '%s'", permProfile.Kind)
```
```suggestion
							return fmt.Errorf("`administrator_account` block is not required when `permission_profile` kind is `%s`", permProfile.Kind)
```

QixiaLu replied:
updated

magodo commented on internal/services/manageddevopspools/managed_devops_pool_resource.go:
```diff
@@ -0,0 +1,679 @@
+									Required:     true,
+									ValidateFunc: validation.StringInSlice(pools.PossibleValuesForAzureDevOpsPermissionType(), false),
+								},
+								"administrator_accounts": {
```
We shall rename it to be singular: `administrator_account`

QixiaLu replied:
Updated, thanks

QixiaLu replied:
Updated, thanks

magodo commented on internal/services/manageddevopspools/managed_devops_pool_resource.go:
```diff
@@ -0,0 +1,679 @@
+func validateManualProfileAgentCounts(profileType string, manualProfile ManualResourcePredictionsProfileModel, maxConcurrency int64) error {
+	// Check all_week_schedule
+	if manualProfile.AllWeekSchedule > 0 && manualProfile.AllWeekSchedule > maxConcurrency {
+		return fmt.Errorf("%s all_week_schedule agent count (%d) cannot exceed maximum_concurrency (%d)",
```
```suggestion
		return fmt.Errorf("%s `all_week_schedule` agent count (%d) cannot exceed `maximum_concurrency` (%d)",
```

QixiaLu replied:
updated, cheers!

magodo commented on internal/services/manageddevopspools/managed_devops_pool_resource_test.go:
```diff
@@ -0,0 +1,427 @@
+				check.That(data.ResourceName).ExistsInAzure(r),
+			),
+		},
+		data.ImportStep(),
```
Can we add one more step to test update from `complete` back to `basic`?

QixiaLu replied:
Included this step, thanks

magodo commented on internal/services/manageddevopspools/managed_devops_pool_resource_test.go:
```diff
@@ -0,0 +1,427 @@
+    }
+  }
+
+  stateful_agent_profile {}
```
Minor, but shouldn't this be `stateless_agent_profile`?

QixiaLu replied:
uh, nice catch! updated

magodo commented on internal/services/manageddevopspools/managed_devops_pool_resource_test.go:
```diff
@@ -0,0 +1,427 @@
+    }
+    image {
+      aliases               = ["well known image"]
+      well_known_image_name = "ubuntu-24.04"
```
Can we change this to another value to test it can be updated?

magodo commented on internal/services/manageddevopspools/managed_devops_pool_resource_test.go:
```diff
@@ -0,0 +1,427 @@
+  location            = azurerm_resource_group.test.location
+  resource_group_name = azurerm_resource_group.test.name
+
+  maximum_concurrency            = 1
```
Can we change this to another value to test it can be updated?

QixiaLu replied:
Updated to 2, which is acceptable by the sku

magodo commented on internal/services/manageddevopspools/models.go:
```diff
@@ -0,0 +1,108 @@
+
+type ResourcePredictionsProfileModel struct {
+	Kind                 string  `tfschema:"kind"`
+	PredictionPreference *string `tfschema:"prediction_preference"`
```
Please avoid using pointer in the model

magodo commented on internal/services/manageddevopspools/models.go:
```diff
@@ -0,0 +1,108 @@
+	Caching            *string `tfschema:"caching"`
+	DiskSizeGB         *int64  `tfschema:"disk_size_gb"`
+	DriveLetter        *string `tfschema:"drive_letter"`
+	StorageAccountType *string `tfschema:"storage_account_type"`
```
Please avoid using pointer in the model

QixiaLu replied:
removed

magodo commented on internal/services/manageddevopspools/models.go:
```diff
@@ -0,0 +1,108 @@
+
+type OrganizationModel struct {
+	Parallelism *int64    `tfschema:"parallelism"`
+	Projects    *[]string `tfschema:"projects"`
```
Please avoid using pointer in the model

magodo commented on internal/services/manageddevopspools/models.go:
```diff
@@ -0,0 +1,108 @@
+
+type AzureDevOpsAdministratorAccountsModel struct {
+	Groups *[]string `tfschema:"groups"`
+	Users  *[]string `tfschema:"users"`
```
Please avoid using pointer in the model

QixiaLu replied:
removed

QixiaLu replied:
removed

magodo reviewed (changes requested):
I've left some more comments for the resource, once resolved I'll further look at the data source. Thanks!

QixiaLu committed:
Multiple fixes from testing

QixiaLu committed:
Fix test

QixiaLu commented on internal/services/manageddevopspools/managed_devops_pool_resource.go:
```diff
@@ -0,0 +1,679 @@
+												Type:     pluginsdk.TypeSet,
+												Required: true,
+												Elem: &pluginsdk.Schema{
+													Type: pluginsdk.TypeString,
```
I have tested that it only accepts certificate versionless secret id. Added the validation. thanks!

QixiaLu commented on internal/services/manageddevopspools/managed_devops_pool_resource.go:
```diff
@@ -0,0 +1,679 @@
+								"well_known_image_name": {
+									Type:     pluginsdk.TypeString,
+									Optional: true,
+								},
```
I have added string is not empty validation.

Justifications: ([source](https://learn.microsoft.com/en-us/azure/devops/managed-devops-pools/configure-images?view=azure-devops&tabs=arm#azure-pipelines-images))
1. `well_known_image_name` aliases will change over time
2. Users are able to append versions, which is difficult to validate

wuxu92 replied:
a validation is still helpful to validate the value should starts with the given list, even we don't validate the version part?

QixiaLu commented on internal/services/manageddevopspools/managed_devops_pool_resource.go:
```diff
@@ -0,0 +1,679 @@
+							Schema: map[string]*pluginsdk.Schema{
+								"parallelism": {
+									Type:     pluginsdk.TypeInt,
+									Optional: true,
```
Verified in Portal that it needs to be in [1, 20000]. Added validation. Thanks!

QixiaLu commented on internal/services/manageddevopspools/managed_devops_pool_resource.go:
```diff
@@ -0,0 +1,679 @@
+						Type:     pluginsdk.TypeList,
+						Optional: true,
+						MaxItems: 1,
+						Computed: true,
```
Thanks for catching it! It is not O+C, but a typo. I have removed `Computed`

QixiaLu replied:
Thanks for catching it! It is not O+C, but a typo. I have removed Computed

QixiaLu commented on internal/services/manageddevopspools/managed_devops_pool_resource.go:
```diff
@@ -0,0 +1,679 @@
+									Elem: &pluginsdk.Resource{
+										Schema: map[string]*pluginsdk.Schema{
+											"groups": {
+												Type:     pluginsdk.TypeSet,
```
Updated, thanks!

QixiaLu replied:
Updated, thanks!

QixiaLu commented on internal/services/manageddevopspools/managed_devops_pool_resource.go:
```diff
@@ -0,0 +1,679 @@
+		case int64:
+			agentCount = count
+		case float64:
+			agentCount = int64(count)
```
updated, thank!

QixiaLu commented on internal/services/manageddevopspools/managed_devops_pool_resource_test.go:
```diff
@@ -0,0 +1,427 @@
+  resource_group_name = azurerm_resource_group.test.name
+
+  maximum_concurrency            = 1
+  dev_center_project_resource_id = azurerm_dev_center_project.test.id
```
Updated, cheers

QixiaLu commented on internal/services/manageddevopspools/managed_devops_pool_resource_test.go:
```diff
@@ -0,0 +1,427 @@
+      well_known_image_name = "ubuntu-24.04"
+      buffer                = "100"
+    }
+    sku_name = "Standard_D2ads_v5"
```
sure, updated. thanks

QixiaLu replied:
sure, updated. thanks

QixiaLu commented on internal/services/manageddevopspools/managed_devops_pool_resource.go:
```diff
@@ -0,0 +1,679 @@
+												},
+											},
+											"users": {
+												Type:     pluginsdk.TypeSet,
```
Updated to TypeList

QixiaLu committed:
Updates from PR review

QixiaLu committed:
Revert change

QixiaLu reviewed (commented):
Hi @magodo , thanks for reviewing! I have addressed all your comments. And the acctests are passed. Could I get another review please?
<img width="1894" height="914" alt="Screenshot 2025-10-31 093651" src="https://github.com/user-attachments/assets/09248cb3-399e-40a1-926a-7b0e637655ec" />

QixiaLu commented on internal/services/manageddevopspools/managed_devops_pool_resource.go:
```diff
@@ -0,0 +1,679 @@
+								},
+								"resource_id": {
+									Type:     pluginsdk.TypeString,
+									Optional: true,
```
added the validation

QixiaLu commented on internal/services/manageddevopspools/managed_devops_pool_resource.go:
```diff
@@ -0,0 +1,679 @@
+								"url": {
+									Type:         pluginsdk.TypeString,
+									Required:     true,
+									ValidateFunc: validation.IsURLWithHTTPS,
```
Updated the order. thanks!

QixiaLu commented on internal/services/manageddevopspools/models.go:
```diff
@@ -0,0 +1,108 @@
+	Aliases            *[]string `tfschema:"aliases"`
+	Buffer             *string   `tfschema:"buffer"`
+	ResourceId         *string   `tfschema:"resource_id"`
+	WellKnownImageName *string   `tfschema:"well_known_image_name"`
```
removed

QixiaLu commented on internal/services/manageddevopspools/models.go:
```diff
@@ -0,0 +1,108 @@
+
+type SecretsManagementSettingsModel struct {
+	CertificateStoreLocation *string  `tfschema:"certificate_store_location"`
+	CertificateStoreName     *string  `tfschema:"certificate_store_name"`
```
removed

QixiaLu committed:
Fix linter and tests

QixiaLu commented:
Hi @magodo , thanks for reviewing! I have updated the PR based on your review, and fix some issues I have noticed. The acctests are passed. Could I get a rereview please? thank you!
<img width="1894" height="914" alt="Screenshot 2025-10-31 093651" src="https://github.com/user-attachments/assets/1be0d3ad-3e47-48a6-98b8-342a6f9931ec" />

magodo commented on internal/services/manageddevopspools/managed_devops_pool_resource.go:
```diff
@@ -0,0 +1,746 @@
+		return nil
+	}
+
+	vmssFabricSlice := vmssFabricValue.AsValueSlice()
```
This will panic if the `vmssFabricValue` is unknown. Instead, you can call `vmssFabricValue.IsWhollyKnown` and return early if it is false. Otherwise if it is known, you can just do the validation based on `vmssFabricProfiles`.

magodo commented on internal/services/manageddevopspools/managed_devops_pool_data_source.go:
```diff
@@ -0,0 +1,364 @@
+			Type:     pluginsdk.TypeInt,
+			Computed: true,
+		},
+		"provisioning_state": {
```
We don't need this as it is useless for the users.

QixiaLu replied:
Removed `provisioning_state`

magodo commented on internal/services/manageddevopspools/managed_devops_pool_data_source_test.go:
```diff
@@ -0,0 +1,42 @@
+		{
+			Config: r.basic(data),
+			Check: acceptance.ComposeTestCheckFunc(
+				check.That(data.ResourceName).Key("location").HasValue(location.Normalize(data.Locations.Primary)),
```
Can we add more checks on those computed attributes to ensure they are correctly set during `Read()`?

magodo commented on website/docs/d/managed_devops_pool.html.markdown:
```diff
@@ -0,0 +1,232 @@
+
+* `name` - (Required) The name of this Managed DevOps Pool. Changing this forces a new Managed DevOps Pool to be created.
+
+* `resource_group_name` - (Required) The name of the Resource Group where the Managed DevOps Pool exists. Changing this forces a new Managed DevOps Pool to be created.
```
```suggestion
* `resource_group_name` - (Required) The name of the Resource Group where the Managed DevOps Pool exists.
```

QixiaLu replied:
removed the force new def

magodo reviewed (dismissed):
Thanks @QixiaLu! I've left some more comments, please take a look!

QixiaLu committed:
Fix from review and reorder args in documents

QixiaLu commented:
Hi @magodo , I have addressed your above comments. I also updated the documentation - fix the ordering, to make it consistent with the `reference-documentation-standards.md`.  here's a rerun of the data source acctest. Could i get a rereview please? thanks!

<img width="1876" height="151" alt="Screenshot 2025-10-31 134947" src="https://github.com/user-attachments/assets/0d0173da-ca25-4e9a-b0f7-600511e8c924" />

QixiaLu commented:
Hi @Lingkar , thanks for testing this resource. As it's an Azure service-related issue, I have shared it with the service team. Please expect some responses from them. Meanwhile, you could find more details [here](https://developercommunity.azure.com/t/Managed-DevOps-Pools-Feature-Request:/10763366?ftype=idea). If it's urgent, please open a support ticket. Thanks for your understanding!

QixiaLu commented:
Hi @davidkarlsen , thanks for reaching out! This PR is still under internal review. I'll try to get some prioritizations on the review.

wuxu92 commented on internal/services/manageddevopspools/managed_devops_pool_resource.go:
```diff
@@ -0,0 +1,726 @@
+							},
+						},
+					},
+					"permission_profile": {
```
Insert a blank line between top-level schema blocks to align with the provider. This formatting should also apply to other schemas.

```suggestion
					},

					"permission_profile": {
```

wuxu92 commented on internal/services/manageddevopspools/managed_devops_pool_resource.go:
```diff
@@ -0,0 +1,726 @@
+													ValidateFunc: validate.Email,
+												},
+											},
+										},
```
IIRC we should not define all nested arguments as optional, at least one should be required, or `AtLeastOneOf` should be added, so user would not define a empty block in configuration.

wuxu92 commented on internal/services/manageddevopspools/managed_devops_pool_resource.go:
```diff
@@ -0,0 +1,726 @@
+									Type:         pluginsdk.TypeString,
+									Optional:     true,
+									ValidateFunc: validation.StringIsNotEmpty,
+								},
```
can we specify the `aliases`, `resource_id` and `well_known_image_name` at the same time? if not we may add a conflict with to make it clear and easier to use.

QixiaLu replied:
`resource_id` and `well_knwon_image_name` cannot be set at the same time. Because `image` is a TypeList with more than 1 item, I cannot add `ExactlyOneOf`. So I implemented `validateVmssFabricProfileImages` in `CustomizeDiff` as a workaround.

wuxu92 reviewed (commented):
Thank you for the new resources! I've added a few comments; otherwise, everything looks good to me. 🏷️

QixiaLu committed:
Updates from review

QixiaLu committed:
make fmt

QixiaLu committed:
fmt fixes

wuxu92 reviewed (approved):
Thanks @QixiaLu LGTM ☘️

QixiaLu commented:
Hi @sreallymatt , thanks for reviewing this PR🥇! I have addressed your comments above. And our team has done a round of internal review. It's now ready for some rereviews from your end. Cheers!

sreallymatt commented on internal/services/manageddevopspools/managed_devops_pool_data_source.go:
```diff
@@ -0,0 +1,381 @@
+
+		"tags": commonschema.TagsDataSource(),
+
+		"vmss_fabric_profile": {
```
Could you update this following our naming conventions? ([ref](https://github.com/hashicorp/terraform-provider-azurerm/blob/main/contributing/topics/reference-naming.md?plain=1#L15-L16))

```suggestion
		"virtual_machine_scale_set_fabric_profile": {
```

sreallymatt commented on internal/services/manageddevopspools/managed_devops_pool_data_source.go:
```diff
@@ -0,0 +1,381 @@
+												Computed: true,
+											},
+
+											"disk_size_gb": {
```
Could you update this to `disk_size_in_gb` to follow our naming conventions?

```suggestion
											"disk_size_in_gb": {
```

sreallymatt commented on internal/services/manageddevopspools/managed_devops_pool_resource.go:
```diff
@@ -0,0 +1,751 @@
+			},
+		},
+
+		"dev_center_project_resource_id": commonschema.ResourceIDReferenceRequired(&projects.ProjectId{}),
```
When referencing an Azure Resource ID we generally omit `resource` from the field name

```suggestion
		"dev_center_project_id": commonschema.ResourceIDReferenceRequired(&projects.ProjectId{}),
```

sreallymatt commented on internal/services/manageddevopspools/managed_devops_pool_resource.go:
```diff
@@ -0,0 +1,751 @@
+												ValidateFunc: validation.StringInSlice(pools.PossibleValuesForCachingType(), false),
+											},
+
+											"disk_size_gb": {
```
```suggestion
											"disk_size_in_gb": {
```

sreallymatt commented on internal/services/manageddevopspools/managed_devops_pool_resource.go:
```diff
@@ -0,0 +1,751 @@
+				for _, permProfile := range orgProfile.PermissionProfile {
+					if permProfile.Kind != string(pools.AzureDevOpsPermissionTypeSpecificAccounts) {
+						if len(permProfile.AdministratorAccounts) > 0 {
+							return fmt.Errorf("`administrator_account` block is not required when `permission_profile` kind is `%s`", permProfile.Kind)
```
Given we're returning an error and preventing this configuration from being provisioned, would it make sense to update the message to something like the below?

imo `is not required` makes it sound optional, but not necessarily incorrect

```suggestion
							return fmt.Errorf("`administrator_account` block must not be specified when `permission_profile` kind is `%s`", permProfile.Kind)
```

wuxu92 commented on internal/services/manageddevopspools/managed_devops_pool_resource.go:
```diff
@@ -0,0 +1,751 @@
+						Type:         pluginsdk.TypeString,
+						Optional:     true,
+						Default:      "7.00:00:00",
+						ValidateFunc: validation.StringIsNotEmpty,
```
add validate for `Agent lifetime for stateful pools can not be higher than 7 days (07.00:00:00).`

QixiaLu replied:
Added additional validation. Thx!

Currently confirming additional validation format with the service team. In portal, it doesn't accept values with `hh` greater than 7, e.g. `08:00:00`. But API request is successful.

wuxu92 commented on internal/services/manageddevopspools/managed_devops_pool_resource.go:
```diff
@@ -0,0 +1,751 @@
+	}
+
+	for _, vmssFabricProfile := range vmssFabricProfiles {
+		for i, image := range vmssFabricProfile.Images {
```
if the config with a empty `image {}` block, it will not be loaded to `vmssFabricProfile`, so user will see drift after apply  below, but if we add default value to `buffer` field, this issue not exists.

```
      ~ vmss_fabric_profile {
            # (2 unchanged attributes hidden)

          + image {}

            # (1 unchanged block hidden)
        }
```

QixiaLu replied:
Added default value, cheers

wuxu92 reviewed (commented):
I have left some comments to update the validate function and default value of devops pools

wuxu92 commented on internal/services/manageddevopspools/managed_devops_pool_resource.go:
```diff
@@ -0,0 +1,751 @@
+									Type:         pluginsdk.TypeString,
+									Optional:     true,
+									ValidateFunc: validation.StringInSlice(pools.PossibleValuesForOsDiskStorageAccountType(), false),
+								},
```
this property can move `os_disk_storage_account_type` to upper level to the sibling of `sku_name` and `images` just like the Portal, then the `storage_profile` focus on the the data disk and we can simplify the schema by removing the `data_disk` layer. `os_disk_storage_account_type` should have a default value as Standard or the API will response error for config like:

```
    storage_profile {
      data_disk {
      }
    }
```

wuxu92 commented on internal/services/manageddevopspools/managed_devops_pool_resource.go:
```diff
@@ -0,0 +1,751 @@
+
+											"disk_size_gb": {
+												Type:         pluginsdk.TypeInt,
+												Optional:     true,
```
data disk size is required not optional, and to align with the guidance the naming should be `disk_size_in_gb`

wuxu92 commented on internal/services/manageddevopspools/managed_devops_pool_resource.go:
```diff
@@ -0,0 +1,751 @@
+											"certificate_store_name": {
+												Type:         pluginsdk.TypeString,
+												Optional:     true,
+												ValidateFunc: validation.StringInSlice(pools.PossibleValuesForCertificateStoreNameOption(), false),
```
though the swagger and SDK defined this as an enum list and the possible values, but "My" and "Root" are actually only for windows, for linux this name can be any possible directory, we need to confirm with service team if the swagger need to be updated. if yes then we have to support other values rather than only "My" and "Root".

QixiaLu replied:
Confirming with the service team, will update once they get back on this. thx!

wuxu92 commented on internal/services/manageddevopspools/managed_devops_pool_resource.go:
```diff
@@ -0,0 +1,751 @@
+			},
+		},
+
+		"stateful_agent_profile": {
```
can we rename all properties `xxx_profile` by removing suffix `_profile` according to guidance https://github.com/hashicorp/terraform-provider-azurerm/blob/7a5c692e36666cbcc8459b53b260a6ad89bdb915/contributing/topics/reference-naming.md?plain=1#L18 .

QixiaLu replied:
Removed `_profile` suffix

wuxu92 reviewed (commented):
this resource works over all, I left some comments to improve the schema design, validation and behavior align with Portal.

wuxu92 commented on internal/services/manageddevopspools/managed_devops_pool_resource.go:
```diff
@@ -0,0 +1,751 @@
+											"key_export_enabled": {
+												Type:     pluginsdk.TypeBool,
+												Required: true,
+											},
```
defaults to `false`?

wuxu92 commented on internal/services/manageddevopspools/managed_devops_pool_resource.go:
```diff
@@ -0,0 +1,751 @@
+												Required: true,
+											},
+
+											"observed_certificates": {
```
rename to `key_vaule_certificate_url` for clear and align with other resource of provider

QixiaLu replied:
Updated, thx!

wuxu92 commented on internal/services/manageddevopspools/managed_devops_pool_resource.go:
```diff
@@ -0,0 +1,751 @@
+									ValidateFunc: validation.StringInSlice(pools.PossibleValuesForLogonType(), false),
+								},
+
+								"secrets_management": {
```
rename it to liek `key_vault_management`

wuxu92 commented on internal/services/manageddevopspools/helpers.go:
```diff
@@ -0,0 +1,747 @@
+				},
+				"monday_schedule": {
+					Type:          pluginsdk.TypeMap,
+					Optional:      true,
```
update the typeMap to a list of block instead of a new list of block above

QixiaLu replied:
Updated to the suggested format and added one more validation:

Same day schedule entry with the same time slot should be merged into a single unit, otherwise, the API will merge them automatically and cause drift:

Config:
```
manual_resource_prediction {
      time_zone_name = "UTC-11"
      monday_schedule {
        count = 2
        time = "02:00:00"
      }
      monday_schedule {
        count = 0
        time = "02:30:00"
      }
      monday_schedule {
        count = 1
        time = "02:30:00"
      }
      monday_schedule {
        count = 0
        time = "05:00:00"
      }
    }
```
API returns:
`"daysData":[{},{"02:00:00":2,"02:30:00":1,"05:00:00":0},{},{},{},{},{}]}`

wuxu92 commented on internal/services/manageddevopspools/helpers.go:
```diff
@@ -0,0 +1,747 @@
+				},
+			},
+		},
+	}
```
this block is fine as the prediction_preference has a default value.

QixiaLu committed:
update schema name to make it consistent with conventions

QixiaLu committed:
fmt

QixiaLu committed:
fmt

QixiaLu committed:
Update schema based on review

QixiaLu committed:
lint fix

QixiaLu committed:
Update properties

QixiaLu committed:
Merge branch 'hashicorp:main' into ajaykn/support-devopsinfrastructure

QixiaLu committed:
Update validation

QixiaLu committed:
Update schema

QixiaLu committed:
Update validation

QixiaLu commented on internal/services/manageddevopspools/managed_devops_pool_resource.go:
```diff
@@ -0,0 +1,751 @@
+						ValidateFunc: commonids.ValidateSubnetID,
+					},
+
+					"os_profile": {
```
Updated to `security`. Thanks!

QixiaLu commented on internal/services/manageddevopspools/managed_devops_pool_resource.go:
```diff
@@ -0,0 +1,751 @@
+
+								"buffer": {
+									Type:     pluginsdk.TypeString,
+									Optional: true,
```
Added default value to buffer, thanks

QixiaLu commented on internal/services/manageddevopspools/managed_devops_pool_resource.go:
```diff
@@ -0,0 +1,751 @@
+						MaxItems: 1,
+						Elem: &pluginsdk.Resource{
+							Schema: map[string]*pluginsdk.Schema{
+								"logon_type": {
```
Updated to boolean type with false as default

QixiaLu commented on internal/services/manageddevopspools/managed_devops_pool_resource.go:
```diff
@@ -0,0 +1,751 @@
+												Required: true,
+												Elem: &pluginsdk.Schema{
+													Type:         pluginsdk.TypeString,
+													ValidateFunc: keyvaultvalidate.NestedItemIdWithOptionalVersion,
```
Updated validation for versionless with Certificate only

QixiaLu committed:
Added additional validation and updated documentation

QixiaLu commented on internal/services/manageddevopspools/managed_devops_pool_resource.go:
```diff
@@ -0,0 +1,751 @@
+									Type:         pluginsdk.TypeInt,
+									Optional:     true,
+									ValidateFunc: validation.IntBetween(1, 10000),
+								},
```
Service team has confirmed that the `parallelism` is an optional field. 

- If users don't provide it, the backend will compute a value for it (could be value other than `0`).
- If users provide a value, it has to be within [1, 10k], and the sum of parallelism equals to the maximum Concurrency or not (so I have added a precheck in the customDiff for this)

QixiaLu committed:
Update docs and validation

QixiaLu commented:
Thanks @wuxu92  for reviewing! I have addressed your comments above and left some follow ups. Could I get a rereview please? Below is the latest acctest results:

<img width="2332" height="899" alt="image" src="https://github.com/user-attachments/assets/6bd1dff5-d08d-481d-a3a2-89e3e9b7d89d" />

QixiaLu committed:
Add workaround for API issues

QixiaLu committed:
Fix test

QixiaLu committed:
Fix test

QixiaLu committed:
Update documentation

QixiaLu committed:
Simplify if else block

QixiaLu committed:
Keep schedule schema as list

QixiaLu committed:
Simplify validation

QixiaLu committed:
Update kind to optional

QixiaLu committed:
Merge branch 'hashicorp:main' into ajaykn/support-devopsinfrastructure

QixiaLu committed:
Remove identity workaround

QixiaLu committed:
Update permission to be force new, as these can only be set during creation

QixiaLu committed:
Update daily schedule to TypeSet

QixiaLu committed:
Workaround for service issue

wuxu92 reviewed (commented):
Thanks @QixiaLu for the updates, it look good to me now, but we can improve the validation function for the name of the resource/data source, once this is fixed we can make this PR ready for review!

QixiaLu committed:
Update validation

QixiaLu committed:
fmt

wuxu92 reviewed (approved):
Thanks @QixiaLu LGTM :shipit:

ms-henglu reviewed (approved):
LGTM

QixiaLu committed:
Upgrade to 2025-09-20

sreallymatt commented on internal/services/manageddevopspools/managed_devops_pool_resource.go:
```diff
@@ -0,0 +1,769 @@
+						ValidateFunc: validate.AgentLifetime,
+					},
+
+					"max_agent_lifetime": {
```
```suggestion
					"maximum_agent_lifetime": {
```

sreallymatt reviewed (changes requested):
Thanks for the work you've put into this PR @QixiaLu - I've left just a couple comments regarding some of our documented conventions. Once those are resolved this should be good to go :shipit:

QixiaLu committed:
Add support for work_folder

QixiaLu committed:
Update based on review

QixiaLu commented on internal/services/manageddevopspools/managed_devops_pool_data_source.go:
```diff
@@ -0,0 +1,383 @@
+			},
+		},
+
+		"vmss_fabric": {
```
Updated to `virtual_machine_scale_set_fabric`. Cheers!

QixiaLu committed:
fmt

QixiaLu committed:
nit: simplify code

QixiaLu committed:
fmt

sreallymatt reviewed (approved):
Thanks @QixiaLu - LGTM 🚢
